system/lib/wasmfs/file.h

```diff
@@ -82,13 +69,17 @@ class File : public std::enable_shared_from_this<File> {
 
 class DataFile : public File {
 
-  virtual __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) = 0;
-  virtual __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) = 0;
+  virtual __wasi_errno_t
+  read(uint8_t* buf, __wasi_size_t len, size_t offset) = 0;
```

Comment:
I think the offset is the offset in the file. That means it can be larger than the size of memory, as we discussed in the meeting. So I think it should be `__wasi_filedelta_t` which is 64-bit.

Reply:
I think that is what `off_t` is for isn't it?    But also shouldn't this be called `pread` if you it includes the offset?

Reply:
> But also shouldn't this be called pread if you it includes the offset?

We could change the name, but this is referring to the `read` on the File, not the actual syscall. There is also a `__wasi_fd_pread`. I don't think it would make sense to use `pread` here.